README.md
las2peer-ActivityTracker
===================

Idea
-------------------
las2peer-ActivityTracker is a microservice to provide activity information for las2peer services. The service works together with all kinds of microservices.

The activity tracker response to HTTP or (las2peer) p2p requests and can fetch activity objects from the origin microservice which created an activity.
This service has also build-in MQTT support, so that all new activities will get publish to an MQTT broker of your choice. To enable MQTT publish please provide the MQTT information in the configuration file.

We also provide a modern webcomponent frontend for this service which you can find on **[<i class="icon-link "></i>Github](https://github.com/rwth-acis/activity-tracker)**.

The service is under development. You can participate by creating pull requests or by discussing ideas and requirements inside **[<i class="icon-link "></i>Requirements-Bazaar](https://requirements-bazaar.org/projects/2/categories/169)**.

More information about our microservice ecosystem is explained in our  **[<i class="icon-link "></i>Requirements-Bazaar wiki on Github](https://github.com/rwth-acis/RequirementsBazaar/wiki)**.

----------

Service Documentation
-------------------
We use **[<i class="icon-link "></i>Swagger](http://swagger.io/specification/)** for documenting this microservice. You can use **[<i class="icon-link "></i>Swagger UI](http://swagger.io/swagger-ui/)** to inspect the API.
You can inspect the API documentation of our live (master branch) instance an our beta (develo branch) with our **[<i class="icon-link "></i>Swagger UI instance](http://requirements-bazaar.org/docs/)**.

API documentation endpoint:

 - `baseURL/activities/swagger.json`

----------

Technology
-------------------
The activity tracker is built on Java technologies. As a service framework we use our in-house developed **[<i class="icon-link "></i>las2peer](https://github.com/rwth-acis/LAS2peer)** project.
For persisting our data we use a PostgreSQL database and jOOQ to access it. User input validation is done using Java Bean Validation and for serializing our data into JSON format, we use the Jackson library. As MQTT client we use Eclipse Paho.

Dependencies
-------------------
In order to be able to run this service project the following components should be installed on your system:

 - JDK 14
 - PostgreSQL 12 or newer
 - Gradle to build

----------

How to set up the database
-------------------
 1. `git clone` this repo
 2. To configure your database access look at the [Configuration](#configuration) section
 3. Create a new database called `reqbaztrack`, possibly with UTF-8 collation
 4. Compile the project with `./gradlew build`
 5. Run `./gradlew update` to create your db schema or migrate to a newer version while updating your service
 6. If you need sample data import the file `etc/add_activitytracker_demo_data.sql` into your database.

Configuration
-------------------
You need to configure the service to your own specific environment. Here is the list of configuration variables:

`etc/de.rwth.dbis.acis.activity.service.ActivityService.properties`:
 - `dbUserName`:	Database username, which will be used to access the database
 - `dbPassword`:	Database user password, which will be used to access the database
 - `dbUrl`:			JDBC Connection string to access the database
 - `dbVendor`:      Choose the backend db to use (mysql or postgres (default: postgres))
 - `baseURL`:       Base URL this service runs on
 - `mqttBroker`:    MQTT Broker, if this field is set it enables MQTT publish of new activities (optional).
 - `mqttUserName`:  MQTT username to publish to broker, if this field is set MQTT use username and password. If not it MQTT doe not use authorize to broker (optional).
 - `mqttPassword`:  MQTT password to publish to broker (optional)
 - `mqttOrganization`: Your organisation name, used as first channel description in MQTT (optional)

For other configuration settings, check the **[<i class="icon-link "></i>las2peer](https://github.com/rwth-acis/LAS2peer)** project.

Build
-------------------
For build management we use Gradle. To build the cloned code, please use a console/terminal open the projects root directory and run:

 - `./gradlew build`

Make sure your database settings are configured correctly in the `settings.gradle` and `activity_tracker/etc` setting,  since the test cases require a database.
Liquibase will automatically take care of applying the necessary database schema migrations.

How to run
-------------------
 1. First please make sure you have already [set up the database](#how-to-set-up-the-database)
 2. Make sure your [config settings](#configuration) and gradle settings are properly set.
 3. [Build](#build)
 4. Open a console/terminal window and navigate to the `bin` directory
 5. Run the `start_network.bat` or `start_network.sh` script

How to run using Docker
-------------------

First build the image:
```bash
docker build . -t activity-tracker
```

Then you can run the image like this:

```bash
docker run -e POSTGRES_USER=myuser -e POSTGRES_PASSWORD=mypasswd -p 8080:8080 -p 9011:9011 activity-tracker
```

Replace *myuser* and *mypasswd* with the username and password of a PostgreSQL user with access to a database named *reqbaztrack*.
By default the database host is *postgres* and the port is *5432*.
The REST-API will be available via *http://localhost:8080/activities* and the las2peer node is available via port 9011.

In order to customize your setup you can set further environment variables.

### Node Launcher Variables

Set [las2peer node launcher options](https://github.com/rwth-acis/las2peer-Template-Project/wiki/L2pNodeLauncher-Commands#at-start-up) with these variables.
The las2peer port is fixed at *9011*.

| Variable | Default | Description |
|----------|---------|-------------|
| BOOTSTRAP | unset | Set the --bootstrap option to bootrap with existing nodes. The container will wait for any bootstrap node to be available before continuing. |
| SERVICE_PASSPHRASE | Passphrase | Set the second argument in *startService('<service@version>', '<SERVICE_PASSPHRASE>')*. |
| SERVICE_EXTRA_ARGS | unset | Set additional launcher arguments. Example: ```--observer``` to enable monitoring. |

### Service Variables

See [configuration](#configuration) for a description of the settings.

| Variable | Default |
|----------|---------|
| POSTGRES_USER | *mandatory* |
| POSTGRES_PASSWORD | *mandatory* |
| POSTGRES_HOST | postgres |
| POSTGRES_PORT | 5432 |
| BASE_URL | http://localhost:8080/activities/ |
| MQTT_BROKER | "" |
| MQTT_USER | "" |
| MQTT_PASSWORD | "" |
| MQTT_ORGANIZATION | "" |

### Web Connector Variables

Set [WebConnector properties](https://github.com/rwth-acis/las2peer-Template-Project/wiki/WebConnector-Configuration) with these variables.
*httpPort* and *httpsPort* are fixed at *8080* and *8443*.

| Variable | Default |
|----------|---------|
| START_HTTP | TRUE |
| START_HTTPS | FALSE |
| SSL_KEYSTORE | "" |
| SSL_KEY_PASSWORD | "" |
| CROSS_ORIGIN_RESOURCE_DOMAIN | * |
| CROSS_ORIGIN_RESOURCE_MAX_AGE | 60 |
| ENABLE_CROSS_ORIGIN_RESOURCE_SHARING | TRUE |
| OIDC_PROVIDERS | https://api.learning-layers.eu/o/oauth2,https://accounts.google.com |

### Other Variables

| Variable | Default | Description |
|----------|---------|-------------|
| DEBUG  | unset | Set to any value to get verbose output in the container entrypoint script. |
| INSERT_DEMO_DATA | unset | Set to any value to insert demo data into the database at startup. |

### Custom Node Startup

If the variables are not sufficient for your setup you can customize how the node is started via arguments after the image name.
In this example we start the node in interactive mode:
```bash
docker run -it -e POSTGRES_USER=myuser -e POSTGRES_PASSWORD=mypasswd activity-tracker startService\(\'de.rwth.dbis.acis.activitytracker.service.ActivityTrackerService@0.6.0\', \'Passphrase\'\) startWebConnector interactive
```
Inside the container arguments are placed right behind the launch node command:
```bash
java -cp lib/* i5.las2peer.tools.L2pNodeLauncher -s service -p ${LAS2PEER_PORT} <your args>
```

### Volumes

The following places should be persisted in volumes in productive scenarios:

| Path | Description |
|------|-------------|
| /src/node-storage | Pastry P2P storage. |
| /src/etc/startup | Service agent key pair and passphrase. |
| /src/log | Log files. |

*Do not forget to persist you database data*

----------

Troubleshooting & FAQ
-------------------
 - I get Java encryption errors: Did you install Java Cryptography Extension?
 - I can not run the start script: Check if you have OS permission to run the file.
 - The service does not start: Check if all jars in the lib and service folder are readable.
 - The start script seems broken: Check if the start script has the correct encoding. If you ran the service on Unix use `dos2unix` to change the encoding.
 - To enable MQTT publish of new activities set `mqttBroker` and `mqttOrganization` in property file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Releases prior to v0.6.0 are only documented on
the [GitHub Release Page](https://github.com/rwth-acis/las2peer-ActivityTracker/releases)

## [Unreleased]

## [0.8.0] - 2021-09-08

### Added

- Added new testcases [#43](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/43)
- Add returned data models to swagger annotations [#46](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/46)

### Changed

- Updated dependencies (las2peer 1.1.2 and thereby requiring Java
    14) [#41](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/41)
        , [#43](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/43),
        [#49](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/49)
- Changed buildsystem to gradle [#43](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/43)
- Automatically generate jooq code from migration files at build
  time [#43](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/43)
- Move to multi-staged docker build for slimmer target
  image [#43](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/43)
- Replaced vtor with Java Bean Validation [#44](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/44)
- Remove static code from data classes and generate getter/setters and builder with
  lombok. [#46](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/46)
- Switch default backend database to postgres [#48](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/48)

### Removed

- `additionalObject` filtering capabilities in `GET /activities`
  endpoint [#44](https://github.com/rwth-acis/las2peer-ActivityTracker/pull/44)

## [0.6.0] - 2019-02-19

See GH Releases


docker-entrypoint.sh
#!/usr/bin/env bash

set -e

# print all comands to console if DEBUG is set
if [[ ! -z "${DEBUG}" ]]; then
  set -x
fi

# set some helpful variables
export SERVICE_PROPERTY_FILE='etc/de.rwth.dbis.acis.activitytracker.service.ActivityTrackerService.properties'
export WEB_CONNECTOR_PROPERTY_FILE='etc/i5.las2peer.connectors.webConnector.WebConnector.properties'

export SERVICE_VERSION=$(awk -F "=" '/service.version/ {print $2}' gradle.properties)
export SERVICE_NAME=$(awk -F "=" '/service.name/ {print $2}' gradle.properties)
export SERVICE_CLASS=$(awk -F "=" '/service.class/ {print $2}' gradle.properties)
export SERVICE=${SERVICE_NAME}.${SERVICE_CLASS}@${SERVICE_VERSION}

export POSTGRES_DATABASE='reqbaztrack'

# check mandatory variables
[[ -z "${POSTGRES_USER}" ]] &&
  echo "Mandatory variable POSTGRES_USER is not set. Add -e POSTGRES_USER=myuser to your arguments." && exit 1
[[ -z "${POSTGRES_PASSWORD}" ]] &&
  echo "Mandatory variable POSTGRES_PASSWORD is not set. Add -e POSTGRES_PASSWORD=mypasswd to your arguments." && exit 1

# set defaults for optional service parameters
[[ -z "${POSTGRES_HOST}" ]] && export POSTGRES_HOST='postgres'
[[ -z "${POSTGRES_PORT}" ]] && export POSTGRES_PORT='5432'

[[ -z "${BASE_URL}" ]] && export BASE_URL="http://localhost:${HTTP_PORT}/activities/"
[[ -z "${SERVICE_PASSPHRASE}" ]] && export SERVICE_PASSPHRASE='Passphrase'
[[ -z "${MQTT_BROKER}" ]] && export MQTT_BROKER=''
[[ -z "${MQTT_USER}" ]] && export MQTT_USER=''
[[ -z "${MQTT_PASSWORD}" ]] && export MQTT_PASSWORD=''
[[ -z "${MQTT_ORGANIZATION}" ]] && export MQTT_ORGANIZATION=''

# set defaults for optional web connector parameters
[[ -z "${START_HTTP}" ]] && export START_HTTP='TRUE'
[[ -z "${START_HTTPS}" ]] && export START_HTTPS='FALSE'
[[ -z "${SSL_KEYSTORE}" ]] && export SSL_KEYSTORE=''
[[ -z "${SSL_KEY_PASSWORD}" ]] && export SSL_KEY_PASSWORD=''
[[ -z "${CROSS_ORIGIN_RESOURCE_DOMAIN}" ]] && export CROSS_ORIGIN_RESOURCE_DOMAIN='*'
[[ -z "${CROSS_ORIGIN_RESOURCE_MAX_AGE}" ]] && export CROSS_ORIGIN_RESOURCE_MAX_AGE='60'
[[ -z "${ENABLE_CROSS_ORIGIN_RESOURCE_SHARING}" ]] && export ENABLE_CROSS_ORIGIN_RESOURCE_SHARING='TRUE'
[[ -z "${OIDC_PROVIDERS}" ]] && export OIDC_PROVIDERS='https://api.learning-layers.eu/o/oauth2,https://accounts.google.com'

# configure service properties

function set_in_service_config() {
  sed -i "s?${1}[[:blank:]]*=.*?${1}=${2}?g" ${SERVICE_PROPERTY_FILE}
}
set_in_service_config dbUserName ${POSTGRES_USER}
set_in_service_config dbPassword ${POSTGRES_PASSWORD}
set_in_service_config dbUrl "jdbc:postgresql://${POSTGRES_HOST}:${POSTGRES_PORT}/${POSTGRES_DATABASE}"
set_in_service_config baseURL ${BASE_URL}
set_in_service_config mqttBroker ${MQTT_BROKER}
set_in_service_config mqttUserName ${MQTT_USER}
set_in_service_config mqttPassword ${MQTT_PASSWORD}
set_in_service_config mqttOrganization ${MQTT_ORGANIZATION}

# configure web connector properties

function set_in_web_config() {
  sed -i "s?${1}[[:blank:]]*=.*?${1}=${2}?g" ${WEB_CONNECTOR_PROPERTY_FILE}
}
set_in_web_config httpPort ${HTTP_PORT}
set_in_web_config httpsPort ${HTTPS_PORT}
set_in_web_config startHttp ${START_HTTP}
set_in_web_config startHttps ${START_HTTPS}
set_in_web_config sslKeystore ${SSL_KEYSTORE}
set_in_web_config sslKeyPassword ${SSL_KEY_PASSWORD}
set_in_web_config crossOriginResourceDomain ${CROSS_ORIGIN_RESOURCE_DOMAIN}
set_in_web_config crossOriginResourceMaxAge ${CROSS_ORIGIN_RESOURCE_MAX_AGE}
set_in_web_config enableCrossOriginResourceSharing ${ENABLE_CROSS_ORIGIN_RESOURCE_SHARING}
set_in_web_config oidcProviders ${OIDC_PROVIDERS}

# set pod ip in pastry conf
if [[ ! -z "${POD_IP}" ]]; then
  echo external_address = ${POD_IP}:${LAS2PEER_PORT} > etc/pastry.properties
  echo socket_bindAddress = ${POD_IP} >> etc/pastry.properties
fi

# wait for any bootstrap host to be available
if [[ ! -z "${BOOTSTRAP}" ]]; then
  echo "Waiting for any bootstrap host to become available..."
  for host_port in ${BOOTSTRAP//,/ }; do
    arr_host_port=(${host_port//:/ })
    host=${arr_host_port[0]}
    port=${arr_host_port[1]}
    if { </dev/tcp/${host}/${port}; } 2>/dev/null; then
      echo "${host_port} is available. Continuing..."
      break
    fi
  done
fi

# prevent glob expansion in lib/*
set -f
LAUNCH_COMMAND='java -cp lib/* i5.las2peer.tools.L2pNodeLauncher -s service -p '"${LAS2PEER_PORT} ${SERVICE_EXTRA_ARGS}"
if [[ ! -z "${BOOTSTRAP}" ]]; then
  LAUNCH_COMMAND="${LAUNCH_COMMAND} -b ${BOOTSTRAP}"
fi

# start the service within a las2peer node
if [[ -z "${@}" ]]; then
  exec ${LAUNCH_COMMAND} startService\("'""${SERVICE}""'", "'""${SERVICE_PASSPHRASE}""'"\) startWebConnector
else
  exec ${LAUNCH_COMMAND} ${@}
fi
